### DISCUSSION TITLE
Automatically split the data to compute FSC

### DISCUSSION THREAD

**Discussion by vinith2** [zhonge/cryodrgn#439] [Q&A]
Hello, similar to the back projection method where the data is split into two half-maps during reconstruction — is there an option to split the data similarly so that two neural networks can be trained separately and the FSC between the resulting half-maps can be computed?

**Comment by michal-g** [comment]
Hi, we do not currently support such a feature out-of-the-box, as running two separate neural network models concurrently within commands such as `train_vae` would take a fair bit more engineering that running two concurrent backprojection algorithms.

However, we have included tools intended to make it easier for you to split the data and run the reconstruction command twice manually:
```
cryodrgn_utils select_random 252395 \  # replace this with the size of your dataset
                         --frac 0.5 \  # we want to randomly select a half-set
                         -o indices_a.pkl -s indices_b.pkl  # paths to save the two sets of filtering indices
cryodrgn train_vae particles.128.mrcs --ind indices_a.pkl --ctf ctf.pkl ...
cryodrgn train_vae particles.128.mrcs --ind indices_b.pkl --ctf ctf.pkl ...
```

Let us know if this helps or if you have trouble with the above commands!

**Reply by AsDeadAsADodo** [reply]
Hi Michal, I'm a little confused. Do we need to train two separate models in the current pipeline to calculate the FSC? Or what's the right way to obtain FSC figure in current pipeline?

**Reply by michal-g** [reply]
Yes, if you were to obtain a gold-standard FSC from the `train_vae` reconstruction pipeline, you would need to train two separate half-map `train_vae` models and then use the `fsc` (or `plot_fsc`) command on the output half-map volumes. However, this is complicated by the fact that these heterogeneous models return a landscape of volumes rather than a single volume and so you would need to choose a way of selecting matching volumes to compare from each half-map model. There is not presently one "right" way of doing this, and thus not really one "right" way of calculating FSCs between two such models — we would suggest comparing the volumes produced at the latent-space co-ordinates assigned to a small subset of particle images from the dataset by each of the models, similar as to what was done for Figure 3 and the corresponding Per-Image FSC method in the [original cryoDRGN manuscript](https://www.nature.com/articles/s41592-020-01049-4).

Compare this to the case of using the `train_nn` reconstruction pipeline, which uses a homogeneous model, and thus produces a single output volume which can be compared directly across different models, and also the `backproject_voxel` reconstruction pipeline, which too is homogeneous but more importantly much simpler computationally and can therefore train the half-map models as it is training the full model. This also means that `backproject_voxel` can run FSC calculations and produce FSC plots at the conclusion of training, a feature that we currently do not — indeed, cannot — support in our neural network reconstruction pipelines such as `train_vae` and `train_nn`.

Let us know if this clarifies things or if you have any other questions!

**Reply by AsDeadAsADodo** [reply]
Hi Michal,
Thanks for the detailed reply, it really helps!
Regarding Figure 3d, does the "GT density volume" refer to a consensus reconstruction?
Also, is there an online document that introduces the backproject_voxel reconstruction method?

**Reply by michal-g** [reply]
1) Per-image FSCs in this paper were only calculated using results on simulated datasets for which there was a "ground-truth" reconstruction (or rather, distribution of reconstructions) manually constructed as part of the simulation process.
2) Assuming you mean an introduction of the method in general and not our implementation of it specifically (a basic intro of which can be accessed via `cryodrgn backproject_voxel -h`), voxel-based backprojection and its relationship to other reconstruction methods has a pretty good write-up here: https://pmc.ncbi.nlm.nih.gov/articles/PMC3165033/

**Reply by AsDeadAsADodo** [reply]
Thanks, that was really helpful!

**Comment by vinith2** [comment]
Hi,

Thanks, that was very helpful. I tried it out using the EMPIAR-10028 dataset. I performed homogeneous reconstruction in CryoSPARC to obtain the viewing angles and shifts, which I then used to train two models with the following commands:

```
cryodrgn train_nn particles.360.mrcs --poses poses.pkl --ctf ctf.pkl -o output_a/train_nn -n 30 --ind indices_a.pkl --batch-size 20
cryodrgn train_nn particles.360.mrcs --poses poses.pkl --ctf ctf.pkl -o output_b/train_nn -n 30 --ind indices_b.pkl --batch-size 20
```
To compute the FSC, I used this script:

```
from cryodrgn.commands_utils.fsc import calculate_cryosparc_fscs
import mrcfile
import torch

device = 'cpu'

vol_a = './output_a/train_nn/reconstruct.mrc'
vol_a= mrcfile.open(vol_a).data
vol_a = torch.tensor(vol_a, dtype = torch.float32, device = device)

vol_b = './output_b/train_nn/reconstruct.mrc'
vol_b= mrcfile.open(vol_b).data
vol_b = torch.tensor(vol_b, dtype = torch.float32, device = device)
vol_avg = (vol_a + vol_b)/2
Apix = 1.34

calculate_cryosparc_fscs(vol_avg,vol_a,vol_b,apix=Apix,out_file='./fsc/fsc.txt',
                     plot_file='./fsc/fsc.png',
            )

```
Here’s the FSC curve I obtained:

![fsc_standard](https://github.com/user-attachments/assets/df528ba5-bb41-4477-bb29-142d2f6dba8d)

Do you have any idea why the corrected FSC curve drops off so sharply?

**Comment by michal-g** [comment]
https://guide.cryosparc.com/processing-data/tutorials-and-case-studies/tutorial-common-cryosparc-plots#high-resolution-phase-randomization

tl;dr, the "Tight" mask in this case looks to be too tight — i.e. overfitting — and you should try again using a looser set of masks, e.g.:
```
calculate_cryosparc_fscs(
    vol_avg, vol_a, vol_b, apix=Apix, out_file='./fsc/fsc.txt', plot_file='./fsc/fsc.png',
    loose_mask=(40, 20), tight_mask=(25, 15),
)
```

Note that the defaults are `loose_mask=(25, 15), tight_mask=(6, 6)` — that is, the default masks used by CryoSPARC when generating these types of plots, with the parameter pairs representing the size of the cosine dilation and edge distance in Angstrom units respectively.
